### DISCUSSION TITLE
quais são as aplicacões em linguagens de programação?

### DISCUSSION THREAD

**Discussion by camila1467** [frontendbr/forum#1755] [Perguntas e Respostas]
pessoal, eu iniciei o curso de ciência da computação mas nunca tinha programado, dai tudo bem, aprendi a lógica de programação com linguagem c, mas a minha duvida é, se fala muito em PROGRAMAS, sistemas, vaaaarias linguagens mas dai você procura uma aplicação feita pra se ter uma ideia e não acha, joga no yt pra sanar a duvida e só aparece video-aulas, acho que o grande problema nesse assunto é esse, tem muuuuita gente ensinando mas não dizem o básico, nao mostram o caminho a se percorrer, eu iria começar a aprender sozinha java achando que conseguiria criar um programa dai descobri que é só sistema... Por exemplo,o que da pra fazer com java? e python? Não se acha nada que de pra ter uma base entende, voce fica completamente perdido... eu queria começar a fazer programas mas que linguagem uso? python?c#?Eu não sei se entenderam minha perguntas mas, eu queria exemplos praticos e conhecidos feitos em linguagens conhecidas, tipo, photoshop criado em tal linguagem, isso já me daria uma luz no que eu quero de verdade sem da cabeçada

**Comment by felipe-heredia** [comment]
Não sei se te ajudarei bastante mas vamos lá.

Comecei também da mesma forma (ainda estou no inicio), comecei em JavaScript mas não fiz lógica, programação orientada a objetos. Nadaa. _Não me sinto tão perdido, mas aos poucos vou aprendendo mais o que eu não aprendi antes_.

Vou listar alguns sites e aplicativos que sei que foram feitos em React ou ReactNative (JavaScript).

> __Facebook:__ Feito em ReactJS (web) e ReactNative (mobile).
> __Instagram__: Feito em ReactJS (web) e ReactNative (mobile).
> __Airbnb__: Feito em ReactJS.
> __Dropbox__: Feito em ReactJS.

### Enfim, não sei se te ajudei bastante. Mas espero ter ajudado em alguma coisa.

**Comment by camila1467** [comment]
então @felipesuri a parte de front end, usar js eu entendo um pouco, só não entendo as aplicações e projetos criados com demais linguagens, tipo java, python...

**Comment by ir4nbala** [comment]
No caso teria que ver qual ambiente você quer desenvolver, desktop, web,
mobile e etc. Depois ver quais tecnologias o mercado procura, porque não
adianta você dominar linguagem X e o mercado pedir Y. Infelizmente
faculdade não acompanha o mercado, então você teria que investir mais um
pouco de estudo por fora... No meu caso tou investindo em reactjs/rn por
ser bastante procurado, caso se interesse já é um caminho.

**Comment by victorferreira** [comment]
Camila, teoricamente dá pra fazer qualquer coisa com qualquer linguagem, desde que ela seja _Turing complete_, mas aqui vão alguns pontos que eu acho que são importantes pra você ter em mente quando pensar em linguagens de programação.

1. Linguagens de programação são apenas ferramentas. E como qualquer ferramenta, funciona bem pra algumas coisas e não tão bem pra outras. Python pode ser muito bom para criar scripts rápidos mas não vai servir se você quiser criar um app mobile, por exemplo(~até dá pra fazer, mas essa é uma outra discussão~). C pode ser uma boa alternativa se você precisa ter acesso direto ao hardware ou precisa manipular alocação de memória manualmente. Tudo depende dos requisitos da sua aplicação e de quais problemas você está tentando resolver.

2. Provavelmente mais de uma linguagem vai ser capaz de resolver o seu problema, e muitas vezes a escolha de qual linguagem usar é apenas uma questão de preferencia pessoal ou com qual ferramenta o time sente mais confortável. E quanto maior o sistema, ou conjunto de sistemas, maiores as chances de que mais de uma linguagem seja utilizada durante o desenvolvimento.

3. Use a(s) linguagem(s) exigidas pela Universidade para encontrar grupos de usuários e fóruns online e observer as dúvidas postadas nesses fóruns. Lá vai ficar bem claro para que as pessoas estão usando determinada linguagem e quais são as maiores frustrações com a linguagem no dia a dia.

4. Leia sobre a história da linguagem que você escolher. Procure entrevistas com os autores da sua linguagem e qual problema eles estavam tentando resolver quando criaram a ferramenta. Sem entender qual problema uma determinada linguagem resolve pode ficar difícil perceber onde uma determinada linguagem brilha e outra vai mal.

5. Aproveite o curso de Ciência da Computação para aprender muito bem os fundamento sobe software, isso vai fazer muita diferença quando você quiser migrar de tecnologia. Algoritmos e Estruturas de Dados podem parecer matérias meio chatas no início, mas no mundo real faz muita diferença aprender muito bem os fundamentos. Paradigmas de programação vai abrir a sua mente para diversas formas de como um mesmo problema pode ser resolvido, e por aí vai...

**Comment by ninetails** [comment]
Ao decorrer do curso vc vai aprender mais, creio eu. Ciência da Computação geralmente foca bem em baixo nível, geralmente fala de compiladores, estruturas de dados...

Não duvidando da capacidade humana mas seria inviável fazer programas alterando o código binário que será lido pelo computador. Para isso vieram as linguagens de programação, para ser essa _interface_ entre um código que um programador possa ler e um código que uma máquina possa ler. Inicialmente as linguagens faziam apenas instruções e cálculos, sei apenas de [Assembly](https://en.wikipedia.org/wiki/Assembly_language) que é assim. Chamados também de linguagem de baixo nível pois é uma linguagem que mais está perto de instruções binárias de máquina.

Mas mesmo com Assembly, ainda era difícil criar programas. Para facilitar e tornar programação mais acessível, linguagens evoluiram, novas surgiram... Abstrações surgem.

Aí surgem os **compiladores**, programas responsáveis por traduzir uma linguagem que possui uma certa abstração [utilizando idéias de linguística](https://en.wikipedia.org/wiki/Abstract_syntax_tree) para transformar o que foi escrito em instruções e posteriormente em código binário. Daqui pra frente já entramos em linguagens de alto nível.

Outras linguagens de programação criam outros tipos de abstrações como, além do compilador também faz com que o código rode em cima de uma "máquina virtual", algo como rodar o código em cima de um outro programa com mais instruções pré programadas que a máquina nativa falando direto com o hardware. Linguagens tais como Java (Java VM), PHP, JS (v8, Chakra, SpiderMonkey...), Python...

E as linguagens foram surgindo conforme necessidade. Antes de começar a pensar no programa, pense onde ele vai rodar, como ele vai rodar e o que vai precisar de recursos. Também não esqueça quem irá desenvolver.

O exemplo mais confuso e atual: vc quer fazer um app para celular. Daí já pipocam as linguagens nativas: Java para Android e Swift/Objective-C para iOS... Por que essas linguagens? O sistema Android é baseado em Android, você pode até tentar fazer algo diferente de Java, mas já tem tudo preparado lá pra Java, por que ir contra? Se vc escolher outra linguagem não sendo a recomendada, vc vai ter problemas de suporte ao programa gerado ou até mesmo ter que alterar o sistema operacional do aparelho em questão para rodar na linguagem desejada. Porém mesmo que vc faça isso, vc não terá permissão para alterar sistema operacional alheio, dos usuários que utilizariam seu aplicativo.

Mas daí você olha pro desenvolvimento. Tem alguém no seu time que saiba uma das linguagens referenciadas? Pensando em alternativas, vc pode utilizar também um WebView que permite fazer um aplicativo utilizando um navegador embedado e aí programar a aplicação utilizando linguagens web pra front. Se seu app não for autossuficiente, precisará de um servidor para persistir e compartilhar dados. Da mesma forma, escolher uma linguagem de programação que te dê facilidades e agilidade para escrever algo pra web... e por aí vai.

---

Já para saber em qual linguagem é feito certo programa, aplicativo ou sistema, diria que se é uma aplicação proprietária, ou seja, código não aberto e licença apenas para uso, dificilmente vc vai achar qual (ou quais) linguagens pois quem ganhar dinheiro em cima do produto não vai querer que outros copiem ou mesmo deixar público que a aplicação utiliza uma certa linguagem ou bibliotecas de código que podem ter vulnerabilidades onde terceiros podem explorá-las.

O que é diferente se for alguma aplicação/sistema/programa de código aberto. Neste caso, creio que o melhor lugar a aprender é o próprio GitHub (Bitbucket, GitLab....) pois quando o repositório é público, vc pode olhar quais tecnologias foram utilizadas e como.

---

Malz o textão mas fique tranquila que é coisa que vc vai aprender no curso. Ou pelo menos deveria. Tenta ver de fazer matérias como história da computação, se tiver. Ou mesmo ler livros, buscar na net sobre...

**Comment by felquis** [comment]
Pense no que vc gostaria de fazer, e tente entender com a internet/comunidades/faculdade o que aprender e explorar para conseguir alcançar este objetivo.

Geralmente quanto mais impossível parece ser o que vc sonha alcançar, mais interessante e legal fica torna-lo real e mais a gente aprende.
